src/node/internal/internal_timers_promises.ts

```diff
@@ -0,0 +1,218 @@
+  });
+
+  if (signal) {
+    function onCancel(): void {
+      timers.clearImmediate(timer);
+      reject(new AbortError(undefined, { cause: signal?.reason }));
+    }
+    signal.addEventListener('abort', onCancel);
+  }
+
+  return promise;
+}
+
+export async function* setInterval(
+  after?: number,
+  value?: unknown,
+  options: { signal?: AbortSignal; ref?: boolean } = {}
+): AsyncGenerator {
+  if (after !== undefined) {
+    validateNumber(after, 'delay');
+  }
+
+  validateObject(options, 'options');
+
+  if (options.signal !== undefined) {
+    validateAbortSignal(options.signal, 'options.signal');
+  }
+
+  if (options.ref !== undefined) {
+    validateBoolean(options.ref, 'options.ref');
+  }
+
+  // Ref options is a no-op.
+  const { signal, ref } = options;
+
+  if (signal?.aborted) {
+    throw new AbortError(undefined, { cause: signal.reason });
+  }
+
+  let onCancel: (() => void) | undefined;
+  let interval: timers.Timeout;
+  try {
+    let notYielded = 0;
+    let callback: ((promise?: Promise<void>) => void) | undefined;
+    interval = new timers.Timeout(
+      () => {
+        notYielded++;
+        callback?.();
+        callback = undefined;
+      },
+      after,
+      undefined,
+      true,
+      ref
+    );
+
+    if (signal) {
+      onCancel = (): void => {
+        timers.clearInterval(interval);
+        callback?.(
+          Promise.reject(new AbortError(undefined, { cause: signal.reason }))
+        );
+        callback = undefined;
+      };
+      signal.addEventListener('abort', onCancel);
+    }
+
+    while (!signal?.aborted) {
+      if (notYielded === 0) {
+        await new Promise((resolve) => (callback = resolve));
+      }
+      for (; notYielded > 0; notYielded--) {
+        yield value;
+      }
+    }
+    throw new AbortError(undefined, { cause: signal.reason });
+  } finally {
+    // @ts-expect-error TS2454 TS detects invalid use before assignment.
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (interval) {
+      timers.clearInterval(interval);
+    }
+    if (onCancel) {
+      signal?.removeEventListener('abort', onCancel);
+    }
+  }
+}
+
+// TODO(@jasnell): Scheduler is an API currently being discussed by WICG
+// for Web Platform standardization: https://github.com/WICG/scheduling-apis
+// The scheduler.yield() and scheduler.wait() methods correspond roughly to
+// the awaitable setTimeout and setImmediate implementations here. This api
+// should be considered to be experimental until the spec for these are
+// finalized. Note, also, that Scheduler is expected to be defined as a global,
+// but while the API is experimental we shouldn't expose it as such.
+class Scheduler {
+  public [kScheduler] = true;
+
+  public yield(): Promise<void> {
+    if (!this[kScheduler]) throw new ERR_INVALID_THIS('Scheduler');
+    // @ts-expect-error TS2555 Following Node.js implementation
+    return setImmediate();
+  }
+
+  public wait(
```

Comment:
We already implement the global `scheduler.wait(...)` in the runtime. (see `globalThis.scheduler`). This should defer to that instead of being a separate impl.